File: src/main/java/org/openhds/mobile/fragment/navigate/HierarchyButtonFragment.java
package org.openhds.mobile.fragment.navigate;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.LinearLayout.LayoutParams;
import android.widget.RelativeLayout;
import android.widget.ScrollView;

import org.openhds.mobile.R;
import org.openhds.mobile.navconfig.HierarchyPath;
import org.openhds.mobile.navconfig.NavigatorConfig;
import org.openhds.mobile.repository.DataWrapper;

import java.util.HashMap;
import java.util.Map;

import static org.openhds.mobile.utilities.ConfigUtils.getResourceString;
import static org.openhds.mobile.utilities.LayoutUtils.configureTextWithPayload;
import static org.openhds.mobile.utilities.LayoutUtils.makeTextWithPayload;

public class HierarchyButtonFragment extends Fragment implements OnClickListener {

	// for some reason margin in layout XML is ignored
	private static final int BUTTON_MARGIN = 5;

	private ScrollView scrollView;
	private HierarchyButtonListener listener;
	private Map<String, RelativeLayout> levelViews;
	private NavigatorConfig config;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		config = NavigatorConfig.getInstance();
	}

	public interface HierarchyButtonListener {
		void onHierarchyButtonClicked(String level);
	}

	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		if (activity instanceof HierarchyButtonListener) {
			listener = (HierarchyButtonListener)activity;
		}
	}

	@Override
	public void onDetach() {
		super.onDetach();
		listener = null;
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

		View fragmentLayout = inflater.inflate(R.layout.hierarchy_button_fragment, container, false);
		scrollView = (ScrollView)fragmentLayout.findViewById(R.id.hierbutton_scroll);
		ViewGroup buttonLayout = (ViewGroup) fragmentLayout.findViewById(R.id.hierbutton_layout);

		levelViews = new HashMap<>();

		NavigatorConfig config = NavigatorConfig.getInstance();
		for (String level : config.getLevels()) {
			final String description = null;
			RelativeLayout layout = makeTextWithPayload(getActivity(),
                    getResourceString(getActivity(), config.getLevelLabel(level)), description, level, this,
					buttonLayout, R.drawable.data_selector, null, null,true);
			LayoutParams params = (LayoutParams) layout.getLayoutParams();
			params.setMargins(0, 0, 0, BUTTON_MARGIN);
			levelViews.put(level, layout);
			setVisible(level, false);
			setHighlighted(level, false);
		}

		return fragmentLayout;
	}

	public void update(HierarchyPath path) {

		// Configure and show all hierarchy buttons for levels in path
		for (String lvl : path.getLevels()) {
			updateButton(lvl, path.get(lvl));
			setVisible(lvl, true);
		}

		// Hide all buttons not in path
		for (int i = path.depth(); i < config.getLevels().size(); i++) {
			setVisible(config.getLevels().get(i), false);
		}

		// If we can go deeper, enable the next level (disabled with the level name)
		if (path.depth() < config.getLevels().size()) {
			String nextLevel = config.getLevels().get(path.depth());
			updateButton(nextLevel, path.get(nextLevel));
			setVisible(nextLevel, true);
		}

		// Scroll to the bottom when the buttons overflow
		scrollView.post(new Runnable() {
			@Override
			public void run() {
				if (scrollView.canScrollVertically(1)) {
					scrollView.fullScroll(View.FOCUS_DOWN);
				}
			}
		});
	}

	private void updateButton(String level, DataWrapper data) {
		if (data == null) {
			String levelLabel = getString(config.getLevelLabel(level));
			setButtonLabel(level, levelLabel, null, true);
			setHighlighted(level, true);
		} else {
			setButtonLabel(level, data.getName(), data.getExtId(), false);
			setHighlighted(level, false);
		}
	}

	private void setVisible(String level, boolean visible) {
		RelativeLayout layout = levelViews.get(level);
		if (layout != null) {
			layout.setVisibility(visible ? View.VISIBLE : View.GONE);
		}
	}

	private void setHighlighted(String level, final boolean highlighted) {
		final RelativeLayout layout = levelViews.get(level);
		if (layout != null) {
			// Defer setting pressed state so it isn't overwritten when run within a click handler
			new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
				@Override
				public void run() {
					layout.setClickable(!highlighted);
					layout.setPressed(highlighted);
				}
			}, 100);
		}
	}

	private void setButtonLabel(String level, String name, String id, boolean centerText) {
		RelativeLayout layout = levelViews.get(level);
		if (layout != null) {
			configureTextWithPayload(getActivity(), layout, name, id, null, null, centerText);
		}
	}

	@Override
	public void onClick(View v) {
		if (listener != null) {
			listener.onHierarchyButtonClicked((String) v.getTag());
		}
	}
}

File: src/main/kotlin/org/cimsbioko/fragment/navigate/HierarchyButtonFragment.kt
package org.cimsbioko.fragment.navigate

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.ScrollView
import androidx.fragment.app.Fragment
import org.cimsbioko.R
import org.cimsbioko.data.DataWrapper
import org.cimsbioko.navconfig.HierarchyPath
import org.cimsbioko.navconfig.NavigatorConfig
import org.cimsbioko.navconfig.NavigatorConfig.Companion.instance
import org.cimsbioko.utilities.LayoutUtils.configureTextWithPayload
import org.cimsbioko.utilities.LayoutUtils.makeTextWithPayload

class HierarchyButtonFragment : Fragment(), View.OnClickListener {

    private lateinit var scrollView: ScrollView
    private lateinit var levelViews: Map<String, RelativeLayout>
    private lateinit var config: NavigatorConfig

    private var listener: HierarchyButtonListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        config = instance
    }

    interface HierarchyButtonListener {
        fun onHierarchyButtonClicked(level: String)
    }

    override fun onAttach(ctx: Context) {
        super.onAttach(ctx)
        if (ctx is HierarchyButtonListener) listener = ctx
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.hierarchy_button_fragment, container, false).apply {
            scrollView = findViewById(R.id.hierbutton_scroll)
            levelViews = findViewById<ViewGroup>(R.id.hierbutton_layout).let { buttonLayout ->
                val activity = requireActivity()
                config.levels.map { level ->
                    level to makeTextWithPayload(
                            activity, config.getLevelLabel(level), null, level, this@HierarchyButtonFragment,
                            buttonLayout, R.drawable.data_selector, null, null, true).apply {
                        visibility = View.GONE
                        (layoutParams as LinearLayout.LayoutParams).setMargins(0, 0, 0, BUTTON_MARGIN)
                    }
                }.toMap()
            }
        }
    }

    fun update(path: HierarchyPath) {

        // Configure and show all hierarchy buttons for levels in path
        for (lvl in path.levels) {
            updateButton(lvl, path[lvl])
            setVisible(lvl, true)
        }

        // Hide all buttons not in path
        for (i in path.depth() until config.levels.size) {
            setVisible(config.levels[i], false)
        }

        // If we can go deeper, show the next level as disabled with the level name
        if (path.depth() < config.levels.size) {
            config.levels[path.depth()].let { nextLevel ->
                updateButton(nextLevel, path[nextLevel])
                setVisible(nextLevel, true)
            }
        }

        // Scroll to the bottom when the buttons overflow
        scrollView.apply { post { if (canScrollVertically(1)) fullScroll(View.FOCUS_DOWN) } }
    }

    private fun updateButton(level: String, data: DataWrapper?) {
        if (data == null) {
            setButtonLabel(level, config.getLevelLabel(level), null, true)
            setHighlighted(level, true)
        } else {
            setButtonLabel(level, data.name, data.extId, false)
            setHighlighted(level, false)
        }
    }

    private fun setVisible(level: String, visible: Boolean) {
        levelViews[level]?.visibility = if (visible) View.VISIBLE else View.GONE
    }

    private fun setHighlighted(level: String, highlighted: Boolean) {
        levelViews[level]?.let { layout ->
            // Defer setting pressed state so it isn't overwritten when run within a click handler
            Handler(Looper.getMainLooper()).postDelayed({
                layout.isClickable = !highlighted
                layout.isPressed = highlighted
            }, 100)
        }
    }

    private fun setButtonLabel(level: String, name: String?, id: String?, centerText: Boolean) {
        levelViews[level]?.let { layout ->
            configureTextWithPayload(requireActivity(), layout, name, id, null, null, centerText)
        }
    }

    override fun onClick(v: View) {
        listener?.onHierarchyButtonClicked(v.tag as String)
    }

    companion object {
        private const val BUTTON_MARGIN = 5 // margin in layout XML is ignored for some reason
    }
}